File: androidApp/src/main/kotlin/com/mohamedfaridelsherbini/moventiq/ui/onboarding/OnboardingViewModel.kt
package com.mohamedfaridelsherbini.moventiq.ui.onboarding

import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

class OnboardingViewModel(
    private val statusStore: OnboardingStatusStore,
) : ViewModel() {
    private val _state = MutableStateFlow(
        OnboardingUiState(hasCompletedOnboarding = statusStore.hasCompletedOnboarding()),
    )
    val state: StateFlow<OnboardingUiState> = _state.asStateFlow()

    fun onEvent(event: OnboardingEvent) {
        when (event) {
            is OnboardingEvent.PageChanged -> {
                if (event.page !in 0 until OnboardingPage.COUNT) return
                if (event.page == _state.value.currentPage) return
                _state.update { it.copy(currentPage = event.page) }
            }
            OnboardingEvent.Continue -> {
                val page = _state.value.currentPage
                if (page < OnboardingPage.COUNT - 1) {
                    _state.update { it.copy(currentPage = page + 1) }
                } else {
                    finishOnboarding()
                }
            }
            OnboardingEvent.Skip -> finishOnboarding()
        }
    }

    private fun finishOnboarding() {
        statusStore.setOnboardingCompleted()
        _state.update { it.copy(hasCompletedOnboarding = true, isFinished = true) }
    }
}

File: iosApp/iosApp/Features/Onboarding/OnboardingViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class OnboardingViewModel {
    private(set) var state = OnboardingUiState()

    private let statusStore: OnboardingStatusStore

    init(statusStore: OnboardingStatusStore) {
        self.statusStore = statusStore
        state.hasCompletedOnboarding = statusStore.hasCompletedOnboarding()
    }

    func handle(_ event: OnboardingEvent) {
        switch event {
        case let .pageChanged(page):
            guard (0 ..< OnboardingPage.count).contains(page) else { return }
            state.currentPage = page
        case .continue:
            if state.currentPage < OnboardingPage.count - 1 {
                state.currentPage += 1
            } else {
                finishOnboarding()
            }
        case .skip:
            finishOnboarding()
        }
    }

    private func finishOnboarding() {
        statusStore.setOnboardingCompleted()
        state.hasCompletedOnboarding = true
        state.isFinished = true
    }
}

enum OnboardingViewModelFactory {
    @MainActor
    static func makeOnboardingViewModel() -> OnboardingViewModel {
        let arguments = ProcessInfo.processInfo.arguments
        if arguments.contains("-UITestSkipOnboarding") {
            return OnboardingViewModel(statusStore: CompletedOnboardingStatusStore())
        }
        if arguments.contains("-UITestFreshOnboarding") {
            return OnboardingViewModel(statusStore: FreshOnboardingStatusStore())
        }
        return OnboardingViewModel(statusStore: OnboardingPreferences())
    }
}

private struct CompletedOnboardingStatusStore: OnboardingStatusStore {
    func hasCompletedOnboarding() -> Bool { true }
    func setOnboardingCompleted() {}
}

private final class FreshOnboardingStatusStore: OnboardingStatusStore {
    private var completed = false

    func hasCompletedOnboarding() -> Bool { completed }

    func setOnboardingCompleted() {
        completed = true
    }
}
